navit/thread.c

```diff
@@ -0,0 +1,544 @@
+/**
+ * Navit, a modular navigation system.
+ * Copyright (C) 2005-2019 Navit Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA.
+ */
+
+/** @file
+ *
+ * @brief Abstraction layer for system-specific thread routines.
+ */
+
+#include "thread.h"
+#include <glib.h>
+#include <math.h>
+#ifdef HAVE_API_WIN32_BASE
+#include <windows.h>
+#else
+#include <errno.h>
+#include <sys/time.h>
+#define _GNU_SOURCE
+#endif
+#include "debug.h"
+
+/**
+ * @brief Describes the main function for a thread.
+ */
+struct thread_main_data {
+    int (*main)(void *);       /**< The thread’s main function. */
+    void * data;               /**< The argument for the function. */
+};
+
+#ifdef HAVE_POSIX_THREADS
+
+/**
+ * @brief Encapsulates the data structures for a pthread condition.
+ */
+struct thread_event_pthread {
+    pthread_cond_t *cond;
+    pthread_mutex_t *mutex;
+};
+
+/**
+ * @brief A wrapper around the main thread function.
+ *
+ * This wraps the implementation-neutral main function into a function with the signature expected by the POSIX thread
+ * library.
+ *
+ * @param data Pointer to a `struct thread_main_data` encapsulating the main function and its argument.
+ */
+static void *thread_main_wrapper(void * data) {
+    struct thread_main_data * main_data = (struct thread_main_data *) data;
+    void * ret = (void *) (main_data->main(main_data->data));
+    // FIXME free up `data`
+    return ret;
+}
+
+char * thread_format_error(int error) {
```

Comment:
Codefactor does not like this function. The message is that this is too complex.
I disagree with codefactor on this one. Yes, it is a long function but it contains only one switch/case statement. Changing that will require it to be broken into multiple cases actually increase complexity and reduce clarity of code. @jkoan how should we proceed for this one? Do we have the option to ignore/accept some findings as false positives? How would we do it?

Reply:
We can just ignore it, we don't need to do anything about codefactor. I agree that this function is really simple. One solution which would be possible but doesn't gain anything is a map with the error code as key and the string as message. But this isn't needed

Reply:
Done. I had the option